src/utils/components/DiskModal/components/StorageProfileSettings/VolumeMode.tsx

```diff
@@ -1,58 +1,96 @@
+import { Trans } from 'react-i18next';
+import { Link } from 'react-router-dom-v5-compat';
+import { uniq } from 'lodash';
 
-import {
-  V1beta1StorageSpecAccessModesEnum,
-  V1beta1StorageSpecVolumeModeEnum,
-} from '@kubevirt-ui/kubevirt-api/kubevirt';
+import { V1beta1StorageSpecVolumeModeEnum } from '@kubevirt-ui/kubevirt-api/kubevirt';
+import HelpTextIcon from '@kubevirt-utils/components/HelpTextIcon/HelpTextIcon';
+import { documentationURL } from '@kubevirt-utils/constants/documentation';
 import { useKubevirtTranslation } from '@kubevirt-utils/hooks/useKubevirtTranslation';
+import { ClaimPropertySets } from '@kubevirt-utils/types/storage';
 import { FormGroup, Radio } from '@patternfly/react-core';
 
 import { V1DiskFormState } from '../../utils/types';
-import {
-  ACCESS_MODE_FIELD,
-  STORAGE_CLASS_PROVIDER_FIELD,
-  VOLUME_MODE_FIELD,
-  VOLUMEMODE_FIELDID,
-} from '../utils/constants';
-import { getVolumeModeForProvisioner, VOLUME_MODE_RADIO_OPTIONS } from '../utils/modesMapping';
+import { ACCESS_MODE_FIELD, VOLUME_MODE_FIELD, VOLUMEMODE_FIELDID } from '../utils/constants';
+import { getAccessModesForVolume, VOLUME_MODE_RADIO_OPTIONS } from '../utils/modesMapping';
 
-const VolumeMode: FC = () => {
+import RecommendationLabel from './RecommendationLabel';
+
+import './ApplyStorageProfileSettings.scss';
+type VolumeModeProps = {
+  claimPropertySets: ClaimPropertySets;
+};
+
+const VolumeMode: FC<VolumeModeProps> = ({ claimPropertySets }) => {
   const { t } = useKubevirtTranslation();
 
-  const { control, setValue, watch } = useFormContext<V1DiskFormState>();
+  const { setValue, watch } = useFormContext<V1DiskFormState>();
 
   const volumeMode = watch(VOLUME_MODE_FIELD);
-  const accessModes = watch(ACCESS_MODE_FIELD) as V1beta1StorageSpecAccessModesEnum[];
-  const storageClassProvisioner = watch(STORAGE_CLASS_PROVIDER_FIELD);
+  const recommendedVolumeModes = uniq(
+    claimPropertySets
+      .map((it) => it.volumeMode)
+      .filter((mode) =>
+        Object.values(V1beta1StorageSpecVolumeModeEnum).some((key) => key === mode),
+      ),
+  );
 
-  const allowedVolumeModes = useMemo(
-    () => getVolumeModeForProvisioner(storageClassProvisioner, accessModes?.[0]),
-    [accessModes, storageClassProvisioner],
+  const setBothModes = useCallback(
+    (mode) => {
+      const accessModes = getAccessModesForVolume(claimPropertySets, mode);
+      setValue(VOLUME_MODE_FIELD, mode);
+      setValue(ACCESS_MODE_FIELD, accessModes.length > 0 ? [accessModes[0]] : undefined);
+    },
+    [claimPropertySets, setValue],
   );
 
   useEffect(() => {
-    if (!allowedVolumeModes?.includes(volumeMode as V1beta1StorageSpecVolumeModeEnum)) {
-      setValue(VOLUME_MODE_FIELD, allowedVolumeModes[0]);
+    if (volumeMode || !recommendedVolumeModes?.[0]) {
+      return;
     }
-  }, [allowedVolumeModes, volumeMode, setValue]);
+    const mostRecommended = recommendedVolumeModes[0];
+    setBothModes(mostRecommended);
```

Comment:
something in the loigc here is wrong, useEffect trigger setBotModes and setBothModes trigger useEffect

Reply:
The approach follows the existing pattern - `useEffect` hooks are used to update the form based on dynamically retrieved data. We could change the pattern and/or replace react hook form but this requires follow-up issue. 
As for this particular hook: it is triggered with every volume change but early return works as expected and ensures that the setter is used only if volume is undefined (initial run and then on every storage class change)